### ISSUE TITLE
Vim needs a file writing retry on Windows

### ISSUE THREAD

**Issue by skywind3000** [vim/vim#2860]
As we know, if we open a file for reading on windows, the file cannot be opened for writing in another process. Now I have a bunch of vim 8 dedicated plugins installed:

- [vim-signify](https://github.com/mhinz/vim-signify): Show a svn/git diff using Vim its sign column. 
- [ale](https://github.com/w0rp/ale): Asynchronous Lint Engine
- [gutentags](https://github.com/ludovicchabant/vim-gutentags): A Vim plugin that manages your ctags/gtags files

Those tools are very handy, I can't live without them now.

But vim-signify will run git/svn in background to calculate diff, ale will run cppcheck and gcc for linting in background when text changed. gutentags will run ctags and gtags for symbol indexing when text changed.

There are at least 5 background processes running while editing now, then I always get this:

```text
E505: "hello.cpp" is read-only (add ! to override)
```

When I am trying to save the file. It appears that file is busy because some process is reading the file and any `:w` , `:wall` or `:update` will fail in vim.

It can be reproduced by opening a file in python:

```bash
python -c "fp=open('hello.cpp', 'r');raw_input()"
```

before python exits, use `:w` in vim, you will see:

```text
E505: "hello.cpp" is read-only (add ! to override)
```

`hello.cpp` is not read-only, it is just busy. So, I have a ugly script which added a writing retry on windows, if `:w` failed for E505, sleep and retry 30 times:

```VimL
function! s:WriteFileGuard() 
	let windows = has('win32') || has('win64') || has('win95') 
	let windows = windows || has('win32unix') || has('win16')
	let v:errmsg = ''
	if bufname('%') == ''
		echohl ErrorMsg
		echom "E32: No file name"
		echohl None
		return
	endif
	if windows == 0 || &readonly != 0
		try
			exec 'w'
		catch /.*/
			echohl ErrorMsg
			echom v:exception
			echohl None
		endtry
		return
	endif
	let retry = 30
	while retry > 0
		let retry -= 1
		try
			silent exec 'w'
			echom "'". bufname('%'). "' written"
			break
		catch /^Vim\%((\a\+)\)\=:E505/
			sleep 25m
			" echom "retry"
		catch
			echohl ErrorMsg
			echom v:errmsg
			echohl None
			break
		endtry
	endwhile
	if retry <= 0
		silent exec "w!"
	endif
endfunc

command! -nargs=0 WriteFileGuard call s:WriteFileGuard()
```

And problem is solved.

Maybe more and more windows users have no idea about this, and will be confused about `E505` in the future as long as they use a lot of vim 8 dedicated plugins like me.

So, I suppose that it should be a part of builtin file writing commands like `:w`, `:wall` and `:update` in vim, and take care of file busy problem on windows automatically for users.

the routine can be described as following steps:

1. if not running on windows, use the old write file routines and return.
2. if the file is really a read-only file, prompts error `E45` and return.
3. if the file is not a read-only file but it is busy, try 30 times and sleep 10m in each turn until success.
4. if the file is still busy after 30 times trying, throw an exception.

More and more job-oriented plugins will be created in the future. 

Is it a reasonable point to add a file writing retry on windows ?

**Comment by skywind3000** [comment]
update: one thing I forgot to mention:

I don't know the implementation details in file writing. but `:w!` works when file is busy.

`:w!` will over-write the real read-only files, I can't use it as a replacement of `:w` and map it to my `<leader>w`. But maybe retry is not needed on windows and `:w` can borrow some logic from `:w!` to make the file writable if `readonly` is 0.

**Comment by skywind3000** [comment]
update again: 
Maybe using ':w!' instead of ':w' when readonly is 0 is not a good idea than using retry.

Gtags/Ctags may be indexing the file at the time when some one execute a `:w!` command. and will Ctags/Gtags get a file mixed with half old and half new content ?

File is always writable on Linux, it may be dangerous to write file directly on linux without writing it to a temp file and atomic rename to the real file.

Windows don't have a atomic file replacing api. maybe retry is a good choice and we should add it to linux too. I remember eclipse will queue file saving request in some situation and save it 1-2 seconds later.

Retry is a simple version of queue.

**Comment by brammool** [comment]
> As we know, if we open a file for reading on windows, the file cannot be opened for writing in another process. Now I have a bunch of vim 8 dedicated plugins installed:
> 
> - [vim-signify](https://github.com/mhinz/vim-signify): Show a svn/git diff using Vim its sign column. 
> - [ale](https://github.com/w0rp/ale): Asynchronous Lint Engine
> - [gutentags](https://github.com/ludovicchabant/vim-gutentags): A Vim plugin that manages your ctags/gtags files
> 
> Those tools are very handy, I can't live without them now.
> 
> But vim-signify will run git/svn in background to calculate diff, ale will run cppcheck and gcc for linting in background when text changed. gutentags will run ctags and gtags for symbol indexing when text changed.
> 
> There are at least 5 background processes running while editing now, then I always get this:
> 
> ```text
> E505: "hello.cpp" is read-only (add ! to override)
> ```
> 
> When I am trying to save the file. It appears that file is busy because some process is reading the file and any `:w` , `:wall` or `:update` will fail in vim.
> 
> It can be reproduced by opening a file in python:
> 
> ```bash
> python -c "fp=open('hello.cpp', 'r');raw_input()"
> ```
> 
> before python exits, use `:w` in vim, you will see:
> 
> ```text
> E505: "hello.cpp" is read-only (add ! to override)
> ```
> 
> `hello.cpp` is not read-only, it is just busy. So, I have a ugly script which added a writing retry on windows, if `:w` failed for E505, sleep and retry 30 times:
> 
> ```VimL
> function! s:WriteFileGuard() 
> 	let windows = has('win32') || has('win64') || has('win95') 
> 	let windows = windows || has('win32unix') || has('win16')
> 	let v:errmsg = ''
> 	if bufname('%') == ''
> 		echohl ErrorMsg
> 		echom "E32: No file name"
> 		echohl None
> 		return
> 	endif
> 	if windows == 0 || &readonly != 0
> 		try
> 			exec 'w'
> 		catch /.*/
> 			echohl ErrorMsg
> 			echom v:exception
> 			echohl None
> 		endtry
> 		return
> 	endif
> 	let retry = 30
> 	while retry > 0
> 		let retry -= 1
> 		try
> 			silent exec 'w'
> 			echom "'". bufname('%'). "' written"
> 			break
> 		catch /^Vim\%((\a\+)\)\=:E505/
> 			sleep 25m
> 			" echom "retry"
> 		catch
> 			echohl ErrorMsg
> 			echom v:errmsg
> 			echohl None
> 			break
> 		endtry
> 	endwhile
> 	if retry <= 0
> 		silent exec "w!"
> 	endif
> endfunc
> 
> command! -nargs=0 WriteFileGuard call s:WriteFileGuard()
> ```
> 
> And problem is solved.
> 
> Maybe more and more windows users have no idea about this, and will be confused about `E505` in the future as long as they use a lot of vim 8 dedicated plugins like me.
> 
> So, I suppose that it should be a part of builtin file writing commands like `:w`, `:wall` and `:update` in vim, and take care of file busy problem on windows automatically for users.
> 
> the routine can be described as following steps:
> 
> 1. if not running on windows, use the old write file routines and return.
> 2. if the file is really a read-only file, prompts error `E45` and return.
> 3. if the file is not a read-only file but it is busy, try 30 times and sleep 10m in each turn until success.
> 4. if the file is still busy after 30 times trying, throw an exception.
> 
> More and more job-oriented plugins will be created in the future. 
> 
> Is it a reasonable point to add a file writing retry on windows ?

In my opinion we should only do this if we have a reliable way to detect
the situation that the file is opened by another process.  We could then
wait a short time until the other process releases the file.
Does Windows have an API way for this?


-- 
Computers are not intelligent.  They only think they are.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by skywind3000** [comment]
I believe it’s possible to be detected. Windows Explorer always tell me: The action can’t be completed because this file is open in another program. There must be a errno for this, I will check it later.

**Comment by skywind3000** [comment]
update：FILE/fopen/fwite can write a file even if it is opened by another process. errno can't be detected.

```C
#include <stdio.h>
#include <errno.h>

int main(void)
{
	FILE *fp = fopen("d:/test.txt", "w");
	if (fp == NULL) {
		printf("file is not writable, errno is %d\n", errno);
		return 1;
	}
	else {
		int hr;
		printf("open file successfully\n");
		hr = fprintf(fp, "Hello, world !!\n");
		if (hr < 0) {
			printf("writing error, errno is %d\n", errno);
		}
		else {
			printf("file written\n");
		}
	}
	fclose(fp);
	return 0;
}
```

The file is opened before executing this C program:

```batch
python -c "fp=open('d:/test.txt', 'r');raw_input()"
```

and the C program can be executed successfully:

```text
open file successfully
file written
```

I will check open/write routines on windows later.

**Comment by skywind3000** [comment]
I have tried open/fopen/CreateFile on windows, but the file can still be open for writing while another program is reading the file.

Why does vim tell me E505 if I open the file by python ??

**Comment by skywind3000** [comment]
E505 can be found in fileio.c:

```C
	file_readonly = check_file_readonly(fname, (int)perm);

	if (!forceit && file_readonly)
	{
	    if (vim_strchr(p_cpo, CPO_FWRITE) != NULL)
	    {
		errnum = (char_u *)"E504: ";
		errmsg = (char_u *)_(err_readonly);
	    }
	    else
	    {
		errnum = (char_u *)"E505: ";
		errmsg = (char_u *)_("is read-only (add ! to override)");
	    }
	    goto fail;
	}
```

and `check_file_readonly`:

```C
    int
check_file_readonly(
    char_u	*fname,		/* full path to file */
    int		perm)		/* known permissions on file */
{
#ifndef USE_MCH_ACCESS
    int	    fd = 0;
#endif

    return (
#ifdef USE_MCH_ACCESS
# ifdef UNIX
	(perm & 0222) == 0 ||
# endif
	mch_access((char *)fname, W_OK)
#else
	(fd = mch_open((char *)fname, O_RDWR | O_EXTRA, 0)) < 0
					? TRUE : (close(fd), FALSE)
#endif
	);
}
```

It's weird mch_access uses `CreateFile`, and mch_open uses `open` I have tried both of them.

**Comment by skywind3000** [comment]
I can't figure out why `check_file_readonly` returns true, I have simulated all the possible paths
int `check_file_readonly` in the file below:

```C
#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <windows.h>

int test1(void)
{
	FILE *fp = fopen("d:/test.txt", "w");
	if (fp == NULL) {
		printf("file is not writable, errno is %d\n", errno);
		return 1;
	}
	else {
		int hr;
		printf("open file successfully\n");
		hr = fprintf(fp, "Hello, world !!\n");
		if (hr < 0) {
			printf("writing error, errno is %d\n", errno);
		}
		else {
			printf("file written\n");
		}
	}
	fclose(fp);
	return 0;
}

void test2()
{
	HANDLE hh = CreateFileA("d:/test.txt", GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);	
	if (hh == NULL) {
		printf("error in open file, code=%d\n", (int)GetLastError());
		return;
	}
	printf("Open successfully\n");
}

void test3()
{
	int fd = open("d:/test.txt", O_RDWR|0, 0);	
	if (fd < 0) {
		printf("error: %d\n", errno);
	}
	else {
		printf("ok\n");
	}
}


int main(void)
{
	test3();
	return 0;
}
```

**Comment by k-takata** [comment]
Do the external programs open a file with the [`FILE_SHARE_WRITE`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858.aspx) flag?

**Comment by skywind3000** [comment]
no, I tested open/fopen/CreateFile for reading, all of them can cause an E505

**Comment by skywind3000** [comment]
Previously, I tested with python:

```bash
python -c "fp = open('d:/test.txt', 'r'); raw_input()"
```

Before python returns from `raw_input()` , Vim will detect it as a read only file. 

And I also wrote another test by using: open/fopen/CreateFile, the following tests (test1-test3) will open the file in read mode, and wait for user input before process exits.

```C
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <fcntl.h>
#include <errno.h>

void pause()
{
	char ch;
	printf("press enter to continue ...");
	scanf("%c", &ch);
}

void test1()
{
	FILE *fp = fopen("d:/test.txt", "r");
	if (fp == NULL) {
		printf("open failed: %d\n", errno);
		return;
	}
	printf("open for reading ...\n");
	pause();
}

void test2()
{
	HANDLE hh = CreateFileA("d:/test.txt", GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL);
	if (hh == NULL) {
		printf("open failed, code=%d\n", (int)GetLastError());
		return;
	}
	printf("open for reading ...\n");
	pause();
}

void test3()
{
	int fd = open("d:/test.txt", O_RDONLY, 0);
	if (fd < 0) {
		printf("open failed: %d\n", errno);
		return;
	}
	printf("open for reading ...\n");
	pause();
}

int main(void)
{
        // test1();
        // test2();
	test3();
	return 0;
}
```

Vim's `:w` failed in all the tests above and prompted a "E505", as you see, I haven't use the `FILE_SHARE_WRITE` flag, and I have no idea why `check_file_readonly` returns true.

**Comment by skywind3000** [comment]
update, `check_file_readonly` and fileio.c are not involved, I added some log and found out "readonly" was detected by `check_readonly` in ex_cmds.c:

```C
    static int
check_readonly(int *forceit, buf_T *buf)
{
    stat_T	st;

    /* Handle a file being readonly when the 'readonly' option is set or when
     * the file exists and permissions are read-only.
     * We will send 0777 to check_file_readonly(), as the "perm" variable is
     * important for device checks but not here. */
    if (!*forceit && (buf->b_p_ro
		|| (mch_stat((char *)buf->b_ffname, &st) >= 0
		    && check_file_readonly(buf->b_ffname, 0777))))
    {
#if defined(FEAT_GUI_DIALOG) || defined(FEAT_CON_DIALOG)
	if ((p_confirm || cmdmod.confirm) && buf->b_fname != NULL)
	{
	    char_u	buff[DIALOG_MSG_SIZE];

	    if (buf->b_p_ro)
		dialog_msg(buff, _("'readonly' option is set for \"%s\".\nDo you wish to write anyway?"),
		    buf->b_fname);
	    else
		dialog_msg(buff, _("File permissions of \"%s\" are read-only.\nIt may still be possible to write it.\nDo you wish to try?"),
		    buf->b_fname);

	    if (vim_dialog_yesno(VIM_QUESTION, NULL, buff, 2) == VIM_YES)
	    {
		/* Set forceit, to force the writing of a readonly file */
		*forceit = TRUE;
		return FALSE;
	    }
	    else
		return TRUE;
	}
	else
#endif
	if (buf->b_p_ro)
	    EMSG(_(e_readonly));
	else
	    EMSG2(_("E505: \"%s\" is read-only (add ! to override)"),
		    buf->b_fname);
	return TRUE;
    }

    return FALSE;
}
```

When `buf->b_p_ro` is 0 and file is opened in another process, `E505` was emitted here.

**Comment by skywind3000** [comment]
weird. I am still looking into `static int check_readonly`, 

in the condition test:

```C
    if (!*forceit && (buf->b_p_ro
		|| (mch_stat((char *)buf->b_ffname, &st) >= 0
		    && check_file_readonly(buf->b_ffname, 0777))))
```

`mch_stat((char *)buf->b_ffname, &st)` returns 0
`check_file_readonly(buf->b_ffname, 0777)` returns `-1`

`check_file_readonly()` is expected to return 0/FALSE or 1/TRUE, why does it return a `-1` ????

The implementation is very simple:

```C
    int
check_file_readonly(
    char_u	*fname,		/* full path to file */
    int		perm)		/* known permissions on file */
{
#ifndef USE_MCH_ACCESS
    int	    fd = 0;
#endif

    return (
#ifdef USE_MCH_ACCESS
# ifdef UNIX
	(perm & 0222) == 0 ||
# endif
	mch_access((char *)fname, W_OK)
#else
	(fd = mch_open((char *)fname, O_RDWR | O_EXTRA, 0)) < 0
					? TRUE : (close(fd), FALSE)
#endif
	);
}
```

and I expanded it in the `static int check_readonly()` with some log:

```C
    if (!*forceit && (buf->b_p_ro
		|| (mch_stat((char *)buf->b_ffname, &st) >= 0
		    && check_file_readonly(buf->b_ffname, 0777))))
    {
	{
	    logout("[check_readonly] mch_stat -> %d", mch_stat((char*)buf->b_ffname, &st));
	    logout("[check_readonly] check_file_readonly -> %d", check_file_readonly(buf->b_ffname, 0777));
#ifdef USE_MCH_ACCESS
	    logout("[check_readonly] mch_access=%d", mch_access((char *)buf->b_ffname, W_OK));
#else
	    {
		int fd = mch_open((char *)buf->b_ffname, O_RDWR | O_EXTRA, 0);
		logout("[check_readonly] mch_open fd=%d", fd);
		if (fd >= 0) {
		    int sim_retcode = (close(fd), FALSE);
		    logout("[check_readonly] sim_retcode=%d", sim_retcode);
		}
	    }
#endif
```
log output:

```
[2018-05-04 23:46:17] [check_readonly] mch_stat -> 0
[2018-05-04 23:46:17] [check_readonly] check_file_readonly -> -1
[2018-05-04 23:46:17] [check_readonly] mch_open fd=4
[2018-05-04 23:46:17] [check_readonly] sim_retcode=0
```

see, `check_file_readonly` was supposed to return 0/FALSE or 1/TRUE, but it returned `-1` here.
I simulate a part of `check_file_readonly` in `check_readonly` and found it should return 0:

```
[2018-05-04 23:46:17] [check_readonly] mch_open fd=4
[2018-05-04 23:46:17] [check_readonly] sim_retcode=0
```

but actually it returned `-1`.

A bug in MSVC ????

**Comment by skywind3000** [comment]
It is not a bug of MSVC, it appears that `USE_MCH_ACCESS` is defined in fileio.c , but undefined in ex_cmds.c

**Comment by skywind3000** [comment]
I confirm that `mch_access((char*)fname, W_OK)` returns -1 now, and I will check it tomorrow.

**Comment by skywind3000** [comment]
update: `check_file_readonly`  is called by `check_readonly`. I have expended `check_file_readonly` in `check_readonly` and found that:
- using `mch_access` to detect readonly always returns -1
- using `mch_open` to detect readonly always returns 0

And now, I can reproduce this:

```C
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <windows.h>

#define FILE_NAME  "d:/test2.txt"


int main(void)
{
	// touch the new file
	system("echo 123 >> " FILE_NAME);

	// open for read
	FILE *fp = fopen(FILE_NAME, "r");

	if (fp == NULL) {
		printf("open file failed\n");
		return -1;
	}

	// check writable by "open"
	{
		int fd = open(FILE_NAME, O_CREAT|O_WRONLY, 0);
		if (fd < 0) {
			printf("[open] not writable, code=%d\n", errno);
		}	else {
			printf("[open] writable\n");
			close(fd);
		}
	}

	// check writable by "CreateFile"
	{
		HANDLE hh = CreateFileA(FILE_NAME, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
		if (hh == NULL || hh == INVALID_HANDLE_VALUE) {
			printf("[CreateFile] not writable, code=%d\n", (int)GetLastError());
		}	else {
			printf("[CreateFile] writeable\n");
			CloseHandle(hh);
		}
	}

	return 0;
}
```

output:

```
[open] writable
[CreateFile] not writable, code=32
```

Using `open` (simulate mch_open) returns `writable`, but using `CreateFile` (simulate mch_access) returns `not writable`. and get a error code of 32:

> ERROR_SHARING_VIOLATION
>     32 (0x20)
>     The process cannot access the file because it is being used by another process.

Does someone know why `open` and `CreateFile` behave different ??

**Comment by skywind3000** [comment]
I have found a similar problem about `CreateFile` and `ERROR_SHARING_VIOLATION` in SO:
https://stackoverflow.com/questions/4435268/createfile-fails-with-error-error-sharing-violation

CreateFile has a `` parameter:

```C
HANDLE WINAPI CreateFile(
    LPCTSTR               lpFileName,
    DWORD                 dwDesiredAccess,
    DWORD                 dwShareMode,
    LPSECURITY_ATTRIBUTES lpSecurityAttributes,
    DWORD                 dwCreationDisposition,
    DWORD                 dwFlagsAndAttributes,
    HANDLE                hTemplateFile
);
```

If we leave `dwShareMode` to zero, `mch_access` will get a INVALID_FILE_HANDLE and `GetLastError()` will return a ERROR_SHARING_VIOLATION.

The solution is passing `FILE_SHARE_READ` as `dwShareMode` in `CreateFile`. and make `mch_access` behave same as `mch_open` + `close`.

I have write a test for `FILE_SHARE_READ` to confirm it can really distinguish read-only from ERROR_SHARING_VIOLATION:

```C
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <windows.h>

#define FILENAME1 "d:/file_not_readonly.txt"
#define FILENAME2 "d:/file_readonly.txt"

int check_readonly(const char *filename)
{
	HANDLE hh = CreateFileA(filename, GENERIC_WRITE, 
			FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
	if (hh == NULL || hh == INVALID_HANDLE_VALUE) 
		return 1;
	CloseHandle(hh);
	return 0;
}

int main(void)
{
	// touch two files
	system("echo 123 > " FILENAME1);
	system("echo 456 > " FILENAME2);
	system("attrib -r " FILENAME1);
	system("attrib +r " FILENAME2);

	// check readonly
	printf("check_readonly(%s) = %d\n", FILENAME1, check_readonly(FILENAME1));
	printf("check_readonly(%s) = %d\n", FILENAME2, check_readonly(FILENAME2));

	// open two files for reading
	printf("open(%s): fd=%d\n", FILENAME1, open(FILENAME1, O_RDONLY, 0));
	printf("open(%s): fd=%d\n", FILENAME2, open(FILENAME2, O_RDONLY, 0));

	// check again when files are still opened
	printf("check_readonly(%s) = %d\n", FILENAME1, check_readonly(FILENAME1));
	printf("check_readonly(%s) = %d\n", FILENAME2, check_readonly(FILENAME2));

	return 0;
}
```

output:

```
check_readonly(d:/file_not_readonly.txt) = 0
check_readonly(d:/file_readonly.txt) = 1
open(d:/file_not_readonly.txt): fd=3
open(d:/file_readonly.txt): fd=4
check_readonly(d:/file_not_readonly.txt) = 0
check_readonly(d:/file_readonly.txt) = 1
```

See, no matter the file is opened in another process or not, the new `check_readonly` can really figure out if the file really has a read-only attribute.

**Comment by skywind3000** [comment]
and a patch for this is ready:

This patch will make `mch_access` work exactly like `mch_open` + `close`, and provide a corresponding behavior to `check_file_readonly` between different systems.

```patch
diff --git a/src/os_win32.c b/src/os_win32.c
index c474415c8..757963228 100644
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -6933,14 +6933,15 @@ mch_access(char *n, int p)
     {
        /* Trying to open the file for the required access does ACL, read-only
         * network share, and file attribute checks.  */
+       DWORD sh = ((p & W_OK)? FILE_SHARE_READ : 0) | ((p & R_OK)? FILE_SHARE_WRITE : 0);
        am = ((p & W_OK) ? GENERIC_WRITE : 0)
                | ((p & R_OK) ? GENERIC_READ : 0);
 #ifdef FEAT_MBYTE
        if (wn != NULL)
-           hFile = CreateFileW(wn, am, 0, NULL, OPEN_EXISTING, 0, NULL);
+           hFile = CreateFileW(wn, am, sh, NULL, OPEN_EXISTING, 0, NULL);
        else
 #endif
-           hFile = CreateFile(n, am, 0, NULL, OPEN_EXISTING, 0, NULL);
+           hFile = CreateFile(n, am, sh, NULL, OPEN_EXISTING, 0, NULL);
        if (hFile == INVALID_HANDLE_VALUE)
            goto getout;
        CloseHandle(hFile);
```

If the file doesn't get a `readonly` attribute but is opened in another process, `check_file_readonly` will returns FALSE now, just like on Linux.

And, what's next ??

**Comment by skywind3000** [comment]
I have read the CRT source, `open` / `wopen` will call CreateFile with both FILE_SHARE_READ and FILE_SHARE_WRITE. 

I believe it's safe to add this to mch_access.

**Comment by skywind3000** [comment]
A better patch is ready, after checking CRT source:

```patch
diff --git a/src/os_win32.c b/src/os_win32.c
index c474415c8..ff8c8afb6 100644
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -6931,16 +6931,19 @@ mch_access(char *n, int p)
     }
     else
     {
+       /* Correspond to (mch_open + close) behavior, just like readonly
+        * detection on linux */
+       DWORD sh = FILE_SHARE_READ | FILE_SHARE_WRITE;
        /* Trying to open the file for the required access does ACL, read-only
         * network share, and file attribute checks.  */
        am = ((p & W_OK) ? GENERIC_WRITE : 0)
                | ((p & R_OK) ? GENERIC_READ : 0);
 #ifdef FEAT_MBYTE
        if (wn != NULL)
-           hFile = CreateFileW(wn, am, 0, NULL, OPEN_EXISTING, 0, NULL);
+           hFile = CreateFileW(wn, am, sh, NULL, OPEN_EXISTING, 0, NULL);
        else
 #endif
-           hFile = CreateFile(n, am, 0, NULL, OPEN_EXISTING, 0, NULL);
+           hFile = CreateFile(n, am, sh, NULL, OPEN_EXISTING, 0, NULL);
        if (hFile == INVALID_HANDLE_VALUE)
            goto getout;
        CloseHandle(hFile);
```

**Comment by skywind3000** [comment]
@k-takata , I know you have made a lot of contributions to vim, and very familiar with Win32 development, would you please review this change ?? 

Every windows users will encounter this E505 issue if they use a lot vim8 plugins.

**Comment by skywind3000** [comment]
@brammool k-takata is busy, what to do ?

**Comment by brammool** [comment]
> @brammool k-takata is busy, what to do ?

You mentioned that specific error code that the file appears readonly
because it is in use by another program.  Do we already get that (but
don't use it), or does that require adding a call?
If we already hit it, what is the path from writint a file to there?

-- 
WOMAN:   Well, 'ow did you become king then?
ARTHUR:  The Lady of the Lake, [angels sing] her arm clad in the purest
         shimmering samite, held aloft Excalibur from the bosom of the water
         signifying by Divine Providence that I, Arthur, was to carry
         Excalibur.  [singing stops] That is why I am your king!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by skywind3000** [comment]
You mean do we have already encounter this ? yes.

If not, I will never open an issue here.

TL;DR: it's nothing about `buf_write`, but we need patch `mch_access`, because `mch_access` will treat a in-use file readonly, but actually the file is not readonly. after patch, `E505` disappeared in all the tests and `:w` command works fine even if the file is opened by an external program at the same time.

I have been digging into CRT source and Vim source for three days to get this solution.

**Long Explaination**

I am using gvim on windows, recently I got several "E505: xxxx is read-only (add! to override)" when I executed `:w` command. and it disappeared after another `:w` retry.

The file was not read-only, I know, there must be something wrong, I never got this before, then I checked my `vimrc` , found the `ale` plugin and `gutentags` plugin had been added a few days ago.

And `gutentags` will launch `gtags` program in background on text changed. gtags always take a long time to index symbols from files. According to my habit, I usually save files when I make some modification. And I guess the `E505` error may be caused by saving a file while gtags or linters are reading it.

In order to verify my  suspection, first I open a file in vim (not a read-only file), then use python to open it:

```bash
python -c "fp=open("d:/test.txt", 'r'); raw_input()"
```

The python command will keep the file open until I input something, then switched to vim, use `:w`, as my expectation, I encountered this again:

    E505: "test.txt" is read-only (add ! to override)

So, I guessed it was caused by `buf_write` when file is busy, then wrote a ugly script for writing retry (in my first post), and fired a bug here, supposed that vim need a writing retry on windows.

Later, after reading some functions in `fileio.c` and `os_win32.c` and adding some log in `check_file_readonly()` and `buf_write()`. I realized that it's nothing about `buf_write()` , it is `mch_access()` 's problem:

`mch_access()` uses CreateFile to detect readonly without setting FILE_SHARE_READ and FILE_SHARE_WRITE flags. In some circumstances that linters or gtags is reading the file, `mch_access()` would fail and assume the file is readonly.

But `mch_open` in `buf_write` is fine, because it uses `open` or `_wopen` on Windows, and both `open` and `_wopen` will set the `FILE_SHARE_READ` and `FILE_SHARE_WRITE` flags when they are calling `CreateFile`.

Here is part of `_open` implementation in Visual C++ CRT source:

```C
int __cdecl _topen (
        const _TSCHAR *path,
        int oflag,
        ...
        )
{
        va_list ap;
        int pmode;

        va_start(ap, oflag);
        pmode = va_arg(ap, int);
        va_end(ap);

        /* default sharing mode is DENY NONE */
        return _tsopen(path, oflag, _SH_DENYNO, pmode);
}
```

See that `_SH_DENYNO` is added and it will be interpretered in `_tsopen`:

```C
...
            case _SH_DENYNO:        /* share read and write access */
                fileshare = FILE_SHARE_READ | FILE_SHARE_WRITE;
                break;
...
```

So, it is just a problem of `mch_access()` , `buf_write()` and `mch_open` are not involved. Finally, I added `FILE_SHARE_READ | FILE_SHARE_WRITE` flags to `CreateFile` in `mch_access()` of `os_win32` the issue was solved.

The patched `mch_access()` works fine, and `:w` command never get an `E505` error again. I confirm it with some tests above.

**Comment by ZyX-I** [comment]
I am wondering why bother with checking whether file is readonly at all if it is more logical to just open file and analyze the situation (first errno, do some other checks if errno is not actually sufficient) in order to give proper error message should opening fail?

**Comment by skywind3000** [comment]
No matter whether or not to remove readonly check from `buf_write`, let's first get `mch_access` patched ?

**Comment by micbou** [comment]
I can confirm the patch in https://github.com/vim/vim/issues/2860#issuecomment-386783081 fixes the issue.

**Comment by brammool** [comment]
> I am wondering why bother with checking whether file is readonly at
> all if it is more logical to just open file and analyze the situation
> (first errno, do some other checks if errno is not actually
> sufficient) in order to give proper error message should opening fail?

I don't recall the history, but I believe it was possible to overwrite a
file even though it's marked as read-only.  Not sure what filesystem
that was on.

-- 
Q: How does a UNIX Guru pick up a girl?
A: look; grep; which; eval; nice; uname; talk; date;

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
